### ISSUE TITLE
sequential_distribute_dump fails

### ISSUE THREAD

**Issue by ninnghazad** [GeoscienceAustralia/anuga_core#147]
sequential_distribute_dump fails in cPickle with 

`
File "/usr/local/lib/python2.7/dist-packages/anuga/parallel/sequential_distribute.py", line 246, in sequential_distribute_dump
    cPickle.dump( tostore, f, protocol=cPickle.HIGHEST_PROTOCOL)
SystemError: error return without exception set
`

Probably related:
https://github.com/numpy/numpy/issues/2396 
https://bugs.python.org/issue11564
https://stackoverflow.com/questions/34091717/avoid-cpickle-error-alternatives-to-cpickle-on-python-2-7

As using python3 isn't easily done it seems, i am unable to save large domains. (roughly >100000000 tris) - which is rather annoying when trying multiple variants on same large mesh.

My only advice to other users at this moment would be to not use sequential_distribute_dump to save the combined domain and the parts, but only the partial (per-proc) domains. these *might* just be small enough to work.

As to how to circumvent the error without upgrading anuga to python3, using numpy.save()/numpy.savez() is the most often suggested solution. (not sure if anuga uses numpy arrays for fields internally)

**Comment by stoiver** [comment]
@ninnghazad it wouldn't be too hard to store the information in a netcdf file. Most of the data is either numpy arrays or strings. Maybe boundary_map is a dictionary. 

See line 200 of sequential_distribute.py in anuga/parallel, where the data to be stored is accumulated into an object called tostore, which we then pickle it in the code around line 242 and load around line 272. 

Could make tostore into a class that collects the info and have alternative ways to store and load, ie pickle or netcdf or a combination of numpy save and pickle for the strings.

**Comment by ninnghazad** [comment]
I am currently trying a combination of numpy.save to seperate files and cPickle for the rest. Have to do some bigger tests to see if that is enough for me. However, other parts of the subdivide/dump code seem to not like big meshes either, like pmesh_divide_metis_with_map:
`
sequential_distribute: Subdivide mesh
Error! ***Memory allocation failed for AllocateWorkSpace: edegrees. Requested size: -1415511280 bytes[HOSTNAME:69273] *** Process received signal ***
`
That is with 512G Ram and a lot of disk to spare, so i guess it's some 32bit-limit.

Has anybody successfully done simulations using large (as in > 100Mio tris) meshes before?

**Comment by stoiver** [comment]
@ninnghazad what system are you working on? I am sure we have had a user run with 50M triangles, but I thought he had also gone higher.

**Comment by ninnghazad** [comment]
@stoiver it's some custom numbercruncher, uses two Xeon E5 18c/36t, 512G mem, infiniband and whatnot. it's running debian on a 4.9 kernel and it's a headless system.

my goal is to get to above 500Mio tris to simulate strong rainfall over a medium sized city.

**Comment by stoiver** [comment]
@ninnghazad sounds like an interesting project. It would be great to insure anuga can run on such large problems. You have a large machine, but my guess is that you will need to run such large problems on the nci machine. 

But I am happy to help. An obvious first step would be to change over to metis 5 as suggested in issue #15

**Comment by ninnghazad** [comment]
@stoiver it surely is. ANUGA could be a rather big thing around here, the flexibility as well as being public open source. 

I'll see how goes, but for a start i want to test the whole thing using openmpi without the added hassle of networking and go from there. I assume a simulation working in local parallel is easier to migrate to other nodes than one where i have no idea if the problem is the simulation or the setup.

Anyway, i appreciate you taking the time to talk about this. The original reason for this issue i seem to have circumvented by using numpy's save(), but i'll test more before considering that solved and possibly uploading the rather naive changes i made.

About #15, i actually am trying to shoehorn latest pymetis into distribute_mesh.py. I am still testing, and the way i did it might be too simple, but first results look like it might work. Had to convert domain.neighbours from ndarray to list-copy though, and that might be a rather large amount of not really needed memory-usage. 

Back to testing.

**Comment by stoiver** [comment]
@ninnghazad, I had a look at pymetis, it is probably a good way to go, but an alternative would be to use ctypes with a standard installation of metis 5.1. 

ctypes was not available back when we started the parallel version, back in 2005.

**Comment by ninnghazad** [comment]
<details>
  <summary>some diff</summary>

``````
diff --git a/anuga/parallel/sequential_distribute.py b/anuga/parallel/sequential_distribute.py
index 66939130..35184224 100644
--- a/anuga/parallel/sequential_distribute.py
+++ b/anuga/parallel/sequential_distribute.py
@@ -5,7 +5,6 @@
 
 import numpy as num
 
-
 from anuga import Domain
 
 from anuga.parallel.distribute_mesh  import send_submesh
@@ -217,7 +216,7 @@ def sequential_distribute_dump(domain, numprocs=1, verbose=False, partition_dir=
     """
 
     from os.path import join
-    
+
     partition = Sequential_distribute(domain, verbose, debug, parameters)
 
     partition.distribute(numprocs)
@@ -234,17 +233,25 @@ def sequential_distribute_dump(domain, numprocs=1, verbose=False, partition_dir=
             if exception.errno != errno.EEXIST:
                 raise
 
-    
+    import cPickle
     for p in range(0, numprocs):
 
         tostore = partition.extract_submesh(p) 
 
-        import cPickle
         pickle_name = partition.domain_name + '_P%g_%g.pickle'% (numprocs,p)
         pickle_name = join(partition_dir,pickle_name)
         f = file(pickle_name, 'wb')
-        cPickle.dump( tostore, f, protocol=cPickle.HIGHEST_PROTOCOL)
 
+       lst = list(tostore)
+
+       num.save(pickle_name+".np1",tostore[1],allow_pickle=False) # this append .npy to filename
+       lst[1] = pickle_name+".np1.npy"
+       num.save(pickle_name+".np2",tostore[2],allow_pickle=False)
+       lst[2] = pickle_name+".np2.npy"
+
+       tostore = tuple(lst)
+
+        cPickle.dump( tostore, f, protocol=cPickle.HIGHEST_PROTOCOL)
     return
 
 
@@ -257,7 +264,7 @@ def sequential_distribute_load(filename = 'domain', partition_dir = '.', verbose
 
     pickle_name = filename+'_P%g_%g.pickle'% (numprocs,myid)
     pickle_name = join(partition_dir,pickle_name) 
-    
+
     return sequential_distribute_load_pickle_file(pickle_name, numprocs, verbose = verbose)
 
 
@@ -265,9 +272,9 @@ def sequential_distribute_load_pickle_file(pickle_name, np=1, verbose = False):
     """
     Open pickle files
     """
-    
-    import cPickle    
+
     f = file(pickle_name, 'rb')
+    import cPickle
 
     kwargs, points, vertices, boundary, quantities, boundary_map, \
                    domain_name, domain_dir, domain_store, domain_store_centroids, \
@@ -276,6 +283,9 @@ def sequential_distribute_load_pickle_file(pickle_name, np=1, verbose = False):
                    domain_quantities_to_be_stored, domain_smooth = cPickle.load(f)
     f.close()
 
+    points = num.load(points)
+    vertices = num.load(vertices)
+
     #---------------------------------------------------------------------------
     # Create domain (parallel if np>1)
     #---------------------------------------------------------------------------
``````

</details>

@stoiver this works for me, but it's kinda meh. also didn't look into domain.quantities to see if there are more ndarrays. i would assume that the above would still be problematic if one would not as i do set quantities after the distributed mesh is loaded from cache (so quantities aren't all saved in the cache files, making them smaller). have not yet looked into the save-as-ncdf way you mentioned.

however, a 120mio triangle test worked (see #15). took ~210sec walltime per 1sec simtime on a wet start and with a bit of rain. merging swws at the end of the simulation isn't quite working yet, but sww_merge.py did the job, so it's probably just some usage-error.
as i somewhat expected the merged sww is kind of hard to handle. crayfish won't display it (but gives a proper "std::bad_alloc"), so i will use the to_raster stuff from one of the examples. 

and some stats (1sec yields):
```
Time = 60.0000, delta t in [0.04204673, 0.04808688], steps=22 (207s)
Communication time 1330.96 seconds
Reduction Communication time 1986.94 seconds
Broadcast time 0.00 seconds
Total time 12425.78 seconds
Finished simulation on 70 processors ...
```
that's a lot of comm-overhead right there - suspect my openmpi might be faulty, but that's another issue.

**Comment by stoiver** [comment]
@ninnghazad great work. I'll checkout #15 work. But a comment on your timings. I have found that there is not much advantage to using more than the number of physical cores when running in parallel. So I would suggest trying something like 36 processors and see what timings you get.

**Comment by stoiver** [comment]
@ninnghazad, by the way can you provide some more info on your project. It would be good to point to your projects using these large grids.

**Comment by ninnghazad** [comment]
@stoiver The project just started and will probably take at least a bunch of months to finish. There will be some kind of publication, but i can't yet say if all of the resulting data will be public. It entails multiple simulations for a whole city. I will however make sure ANUGA is prominently named in whatever publications or reports there will be, and get back to you with any material i am allowed to share once the project is further along the timeline. Apart from actual results i am of course happy to share the hows and tips i can give. But there still are ways to go and stuff to figure out before i would advice anybody to use my run.py and setup. Just ran into some problem somewhere above 120 mio triangles when doing:
'''
domain = anuga.shallow_water.shallow_water_domain.Domain(p,v,b)
		
'''
On the other hand converting the 120tri-results to raster (GeoTIFF) for viewing was slow but worked.

**Comment by stoiver** [comment]
@ninnghazad, I was just curious about your project. There is no expectation for you to make the data public. But updates to the code would be much appreciated!

**Comment by ninnghazad** [comment]
@stoiver the goal of the project is to create risk-assessment maps for Dortmund, a city of about 280km2 in western Germany, incorporating local pumping-stations, surface friction as well as different time/rain-sets using a resolution of at most 1m2.

**Comment by ninnghazad** [comment]
@stoiver by the way, updates to the code, do you prefer PRs or diffs?

**Comment by stoiver** [comment]
@ninnghazad our preferred method is PRs. Thanks for the info about your project.

**Comment by Girishchandra-Yendargaye** [comment]
@ninnghazad  Hello Sir...Could you please tell How do you overcome the issues of Large mesh generation and Domain Creation which is sequential step ...We are also trying the same....

**Comment by ninnghazad** [comment]
@Girishchandra-Yendargaye 
I am currently using rectangular meshes, to generate these:
- use a custom tool to generate .json input from large TIF files. creates triangle,point and boundary files.
- use a python script to read these into numpy and pass the arrays to `anuga.shallow_water.shallow_water_domain.Domain(p,v,b)`
- use machines with at least 512GB RAM

This method is not optimal and could easily optimized by for example using numpy's memmap format instead of json as intermediary.

Also remember that depending on what you are simulating, calculating on domains that large will take a *lot* of time. A lot of fast CPUs (with AVX512 preferably) and using properly optimized python/numpy helps. Be prepared to wait weeks for simulations to finish.

**Comment by Girishchandra-Yendargaye** [comment]
> @Girishchandra-Yendargaye
> I am currently using rectangular meshes, to generate these:
> 
> * use a custom tool to generate .json input from large TIF files. creates triangle,point and boundary files.
> * use a python script to read these into numpy and pass the arrays to `anuga.shallow_water.shallow_water_domain.Domain(p,v,b)`
> * use machines with at least 512GB RAM
> 
> This method is not optimal and could easily optimized by for example using numpy's memmap format instead of json as intermediary.
> 
> Also remember that depending on what you are simulating, calculating on domains that large will take a _lot_ of time. A lot of fast CPUs (with AVX512 preferably) and using properly optimized python/numpy helps. Be prepared to wait weeks for simulations to finish.

Can you please provide some sample code... As I dint find such method in shallow water which accepts only triangle,points and boundary Also What about elevation setting in next step which is sequential again...

**Comment by ninnghazad** [comment]
@Girishchandra-Yendargaye
You are correct. Remember that i chose this way because i needed large rectangular meshes - ANUGA's builtin capabilities might provide easier methods to achieve what you may be looking for.

**Comment by Girishchandra-Yendargaye** [comment]
@ninnghazad I am getting below error...What does it mean..

RuntimeError: Error at anuga\abstract_2d_finite_volumes\neighbour_mesh_ext.c:55:
 Segment (336, 337) does not exist

**Comment by ninnghazad** [comment]
i would have to see some data and code to tell you more than the error does. glancing at the source tells me you are trying to construct a boundary with bad parameters.

**Comment by Girishchandra-Yendargaye** [comment]
@ninnghazad Have you tried 12 crores  triangles?
